File: R/height.r
#!/usr/bin/env R
#
# This file is part of CCCA,
# http://github.com/alexjgriffith/CCCA/, 
# and is Copyright (C) University of Ottawa, 2015. It is Licensed under 
# the three-clause BSD License; see LICENSE.txt.
# Author : Alexander Griffith
# Contact: griffitaj@gmail.com

#' loadBedFile
#' 
#' Calls file_length and read_bed from hg19Height.c and returns a 3 column bed file
#' @param file the location of the bed file of interest
#' @return data.frame $chro $start $end
#' @export
#' @template authorTemplate
#' @examples
#'  data<-loadBedFile(file="filelocation")  
loadBedFile<-function(file){
    file<-normalizePath(file)
    if(!file.exists(file)){
        sprintf("Can't find file %s.",file)}
    else{
        fileLength<-as.integer(.C("file_length",file,stringLength=integer(1))[[2]])
        results<-.C("read_bed",file,chro=character(fileLength),start=integer(fileLength),end=integer(fileLength))
        data.frame(chro=results$chro,start=results$start,end=results$end)
    }}

#' convertChroms
#' 
#' Calls valueChromosome from hg19Height.c which converts a chromosme rank to a string
#' @seealso For the reverse see \code{\link{rankChroms}}
#' @param lis a list of chromosmes to
#' @template authorTemplate
#' @export
convertChroms<-function(lis){
    n<-length(lis)
    .C("valueChromosomes",character(n),as.integer(n), lis)[[1]]}

#' @export
getChroms<-function(file,n=100){
    file<-as.character(normalizePath(file))
    if(!file.exists(file)){        
        stop(paste("Can't find file",file))
    }
    else{        
        return(Filter(function(x) x!="", .C("getChroms",file,choms=character(n))$choms))}
}

#' rankChroms
#' 
#' Calls rankChromosome from hg19Height.c which converts a chromosme string to its equivelent rank
#' @seealso For the reverse see \code{\link{convertChroms}}
#' @template authorTemplate
#' @export
rankChroms<-function(lis){
    n<-length(lis)
    .C("rankChromosomes",as.character(lis),as.integer(n), integer(n))[[3]]}

#' getPileUp
#'
#' A minimal wrapper for the pileup function from hg19Height.c
#' returns a integer vector of computed read pileups
#' @param file The raw data file of interest
#' @param bed The preloaded bed infromation including bed$start and bed$end
#' @param chroms a list of chromosomes whos string values have been repalced with ranks
#' @param peakLength The length of the bed data provided
#' @export
getPileUp<-function(file,bed,chroms,peakLength){
    start<-as.integer(as.character(bed$start))
    end<-as.integer(as.character(bed$end))
    peaknum<-as.integer(peakLength)
    score<-integer(peakLength)
    print(sapply(list(start,end,score,chroms),length))
    print(file)
    results<-.C("pileup",file,chrom=chroms,start=start,end=end,peaknum=peaknum,score=score)
    results$score
    #NULL
}

#' PileUp
#'
#' Generates a pile up matrix from a unified set of peaks and a list of raw data sets
#' @template authorTemplate
#' @param data A preloaded bed data.frame which includes slots $chro $start $end
#' @param rawdata a list of raw data files
#' @param n the number of nodes to use. If 0 then the parrallel package is not used
#' @examples
#' # Initialize catagories, files containing raw data 
#' cats<-read.table("/home/griffita/Dropbox/UTX-Alex/jan/catagories")
#' prefix<-"/mnt/brand01-00/mbrand_analysis/data_sets/"
#' suffix<-"_sorted.bed"
#' rawdata<-apply(cats,1,function(x){paste(prefix,x,"/",x,suffix,sep="")})
#' # Apply pileUp to peaks found using MACS
#' data<-hg19Sort(loadBedFile(file))
#' score<-pileUp(data,rawdata,n=22)
#' # cluster the data sets based on the read hights of the peaks
#' temp<-cor(score)
#' rownames(temp)<-t(cats)
#' colnames(temp)<-t(cats)
#' pdf("test.pdf")
#' plot(hclust(dist(temp)),hang=-1)
#' # Rather than using peaks we can do global analyis
#' # this relies on breaking the genome into bins
#' # human.hg19.genome provides the start and stop point for each genome
#' # bin size desides how large the regions should be
#' binSize=10000
#' regions<-do.call(rbind,
#' apply(read.table("/data/binaries/BEDTools/genomes/human.hg19.genome")[1:24,],1,
#' function(x,step) {y<-seq(1,as.numeric(x[2]),step);
#' cbind(as.character(x[1]),as.character(y),as.character(y+step))} ,binSize))
#'   data<-hg19Sort(data.frame(chro=regions[,1],
#'                             start=as.integer(regions[,2]),
#'                             end=as.integer(regions[,3])))
#'   score<-pileUp(data,rawdata,n=22)
#' @export
pileUp<-function(data,rawdata,n=0,verbose=FALSE){
    for(file in rawdata){
        if(!file.exists(file)){
            sprintf("Can't find file %s.",file)
            return}
        if(verbose)
            print(paste("# Raw data file ",file," was found.",sep=""))
    }
    peakLength<-length(data$chro)

    chroms<-rankChroms(data$chro)

    if(n>0){
        cs<-makeForkCluster(n,renice=0)
        ret<-matrix(unlist(parLapply(cs,rawdata,getPileUp,data,chroms,peakLength)),nrow=peakLength)
        stopCluster(cs)
    }else{
    ret<-matrix(unlist(lapply(rawdata,getPileUp,data,chroms,peakLength)),nrow=peakLength)}
    ret}

#' hg19Sort
#'
#' Reorders the chro factor in data to that which is outputed by BWA.
#' This allows for the sorting step to be skipped, however if another chromasome is used then a new order must be defined.
#' 
#' @param data [chr,start,end]
#' @return The same members of data but sorted on chr and start
#' @template authorTemplate
#' @export
hg19Sort<-function(data){    
    neworder<-Filter(function(x) x %in% levels(data$chro), strsplit("chr1 chr2 chr3 chr4 chr5 chr6 chr7 chrX chr8 chr9 chr10 chr11 chr12 chr13 chr14 chr15 chr16 chr17 chr18 chr20 chrY chr19 chr22 chr21 chrM" ," ")[[1]])
    data$chro<-factor(data$chro,neworder)
    data<-data[with(data,order(chro,start)),]
    data
}

#' Chrom Genome Sort
#'
#' Returns a function that reorders the peak data based on the chromosome order provided
#' @param chromOrder A list of chromosomes to be found in the data
#' @examples
#' sortedChroms<-strsplit("chr1 chr2 chr3 chr4 chr5 chr6 chr7 chrX
#' chr8 chr9 chr10 chr11 chr12 chr13 chr14 chr15 chr16 chr17
#' chr18 chr20 chrY chr19 chr22 chr21 chrM" ," ")[[1]]
#' hg19Sort<-chromGenomeSort(sortedChroms)
#' bedData<-loadBedData("test.bed")
#' sortedBedData<-hg19Sort(bedData)
#' @export
chromGenomeSort<-function(chromOrder){
    rf<-function(data){
        ls<-levels(data$chro)
        neworder<-Filter(function(x) {x %in% ls}, chromOrder)
        data$chro<-factor(data$chro,neworder)
        data<-data[with(data,order(chro,start)),]
        data                       
        }
   return(rf)
}
    
#' getData
#'
#' loads data from file returns a sorted set with a height matrix
#' @export
#' @template authorTemplate
#' @param file peak file
#' @param rawdata a list of raw data files
#' @param n the number of nodes used default=0
getData<-function(file,rawdata,n=0){
    data<-hg19Sort(loadBedFile(file))
    score<-pileUp(data,rawdata,n=n)
    c(data=data,score=score)}

File: src/height.c
/* This file is part of CCCA,
   http://github.com/alexjgriffith/CCCA/, 
   and is Copyright (C) University of Ottawa, 2015. It is Licensed under 
   the three-clause BSD License; see LICENSE.txt.
   Author : Alexander Griffith
   Contact: griffitaj@gmail.com */

#include "height.h"


int getChromosome(char value[10])
{
  int i;
  for(i=0;i<25;i++)
    if(!strcmp(value,hg19Chrom[i].value))
      return i;
  return -1;
}

int getChromosomeShort(char value[2])
{
  int i;
  for(i=0;i<25;i++)
    if(value[0]==hg19Chrom[i].shortValue[0] && value[1]==hg19Chrom[i].shortValue[1])
      return i;
  return -1;
}

void getChromosomeValue(int order,char * value)
{
  int i;
  for(i=0;i<25;i++)
    if(order==hg19Chrom[i].order)
      {
	strcpy(value,hg19Chrom[i].value);
	return ;
      }
  strcpy(value,hg19Chrom[25].value);
}

void rankChromosomes(char ** chroms,int *length, int *out)
{
  int i;
  char string[256];
  for(i=0;i<*length;i++)
    {
      sscanf(chroms[i],"%s",string);
      out[i]=getChromosome(string);
    }
}

void valueChromosomes(char ** chroms,int *length, int *out)
{
  int i;
  for(i=0;i<*length;i++)
    {
      getChromosomeValue(out[i],chroms[i]);
    }
}

int compare(int start1, int end1,int start2, int end2)
{
  int width = (end1-start1)*2;
  int difference = (start1+end1-start2-end2);
  if(width-difference<0)
    return 0 ; //move reads forward
  else if(width+difference<0)
    return 1 ; //move peaks forward
  else
    return 2 ; //add to score and move reads forward
}

void pileup(char ** filename,char ** chro,int *start,
	    int *end,int *length,int *scores)
{
#ifdef _VERBOSE
  Rprintf("N=2\tfilename=%s\n",*filename);
#endif
#ifdef _test
  int i;
  for(i=0;i<*length;i++)
    Rprintf("%s\n",chro[i]);
#else
  char  buffer[1024];
  FILE  * f = fopen(*filename,"r");
  char string[1024];
  int inStart,inEnd,inChro;
  int i=0;

  int coll;
  int chrC;
  int ed1C;
  int ed2C;

  while(fgets(buffer,1024, f))
    {      
      sscanf(buffer,"%s\t%d\t%d", string,&inStart,&inEnd);
      chrC=strcmp(chro[i],string); // < 0 next peak
                                   // == 0 next compare
                                   // > 0 next reed

      while(chrC<=0){
	if(i>=*length)
	  break;
	ed1C=inStart-end[i]; //  =< 0 -- inStart < end
	                           //  > 0 next read
	ed2C=start[i]-inEnd; // =< 0-- inEnd > start
                                    // > 0 -- next peak
	if(chrC==0){
	  if(ed1C<=0 && ed2C<=0){
	    scores[i]++;
	    //coll++
	    break;
	    //i++;
	  }
	  else if (ed1C>0){
	    i++;
	  }
	  else{
	    break;
	  }
	    
	}
	else{
	  i++;
	}
	chrC=strcmp(chro[i],string);
	}
      if(i>=*length)
	break;	
    }
  fclose(f);
#endif
}

void file_length(char ** filename,int * i)
{
  char  buffer[1024];
  FILE  * f = fopen(*filename,"r");
  while(fgets(buffer,1024, f))
    {      
      *i=*i+1;
    }
}

void getChroms(char ** filename,char ** chroms)
{

  char  buffer[1024];
  FILE  * f = fopen(*filename,"r");
  char *s;
  char chrom[256];  
  int i=0,start,end;
  int k;  
  int logic=0;
  fgets(buffer,1024, f);
  s = strtok (buffer,"\n");
  sscanf(s,"%s\t%d\t%d",chrom,&start,&end);
  strcpy(chroms[i],chrom);
  i=i+1;
  while(fgets(buffer,1024, f))
    {      
      s = strtok (buffer,"\n");
      sscanf(s,"%s\t%d\t%d",chrom,&start,&end);
      logic=0;
      for(k=0;k<i;k++)
	if( strcmp(chroms[k],chrom)==0)
	  {
	    logic=1;
	    break;
	  }
      if(logic==0)
	{
	  strcpy(chroms[i],chrom);
	  i=i+1;
	}
    
    }
}

void read_bed(char ** filename,char ** chrom,int *start, int *end)
{
  char  buffer[1024];
  FILE  * f = fopen(*filename,"r");
  char  *s;
  char string[256];
  int t;
  int i=0;
  while(fgets(buffer,1024, f))
    {      
      s = strtok (buffer,"\n");
      //sscanf(s,"%s\t%d\t%d",string,&start[i],&end[i]);
      //t=getChromosome(string);
      //getChromosomeValue(t,chrom[i]);
      sscanf(s,"%s\t%d\t%d",chrom[i],&start[i],&end[i]);
      i=i+1;
    }
  fclose(f);
}

